python/packages/core/agent_framework/_workflows/_concurrent.py

```diff
@@ -343,25 +447,92 @@ def build(self) -> Workflow:
         """
-        if not self._participants:
-            raise ValueError("No participants provided. Call .participants([...]) first.")
+        if not self._participants and not self._participant_factories:
+            raise ValueError(
+                "No participants provided. Call .participants([...]) or .register_participants([...]) first."
+            )
 
+        # Internal nodes
         dispatcher = _DispatchToAllParticipants(id="dispatcher")
-        aggregator = self._aggregator or _AggregateAgentConversations(id="aggregator")
+        if isinstance(self._aggregator, Executor):
+            # Case 1: Executor instance - use directly
+            aggregator = self._aggregator
+        elif callable(self._aggregator):
+            # Distinguish between an aggregator factory (could also be a class) and callback-based aggregator
+            if inspect.isclass(self._aggregator):
+                aggregator = self._aggregator()
+            else:
+                # Check the signature: factory has 0 params, callback has 1-2 params
+                sig = inspect.signature(self._aggregator)
+                param_count = len(sig.parameters)
+
+                # Case 2: Executor factory (no parameters) - call it to create the executor
+                # Case 3: Callback with parameters (1-2 params) - wrap in _CallbackAggregator
+                aggregator = self._aggregator() if param_count == 0 else _CallbackAggregator(self._aggregator)  # type: ignore
+
+            if not isinstance(aggregator, Executor):
+                raise TypeError(f"Aggregator factory must return an Executor; got {type(aggregator).__name__}")
+        else:
+            # Case 4: No custom aggregator provided - use the default one
+            aggregator = _AggregateAgentConversations(id="aggregator")
+
+        participants: list[Executor | AgentProtocol] = []
+        if self._participant_factories:
+            for factory in self._participant_factories:
+                p = factory()
+                if not isinstance(p, (AgentProtocol, Executor)):
+                    raise TypeError(
+                        f"Participant factory must return AgentProtocol or Executor; got {type(p).__name__}"
+                    )
+                participants.append(p)
+        else:
+            participants = self._participants
 
         builder = WorkflowBuilder()
-        builder.set_start_executor(dispatcher)
-        builder.add_fan_out_edges(dispatcher, list(self._participants))
-
-        if self._request_info_enabled:
-            # Insert interceptor between fan-in and aggregator
-            # participants -> fan-in -> interceptor -> aggregator
-            request_info_interceptor = RequestInfoInterceptor(executor_id="request_info")
-            builder.add_fan_in_edges(list(self._participants), request_info_interceptor)
-            builder.add_edge(request_info_interceptor, aggregator)
+        if self._participant_factories:
```

Comment:
Can we resolve factories to instances first, then use a single wiring path? The fan-out/fan-in logic below is duplicated between the two branches.